### PR TITLE
Bespoke smirks generation fix

### LINKED ISSUE #40: "IndexError: list index out of range" when no fragments are generated by fragmentation_engine
Building a schema with molecules like ethanol failed at  

```

from openff.bespokefit.workflows import bespoke
workflow = bespoke.BespokeWorkflowFactory(
    initial_force_field="openff_unconstrained-1.3.0.offxml", 
)

from openff.toolkit.topology import Molecule 
eth_mol = Molecule.from_smiles("CCO")
schema = workflow.optimization_schema_from_molecule(molecule=eth_mol)

```
with the following error: 

```
bespoke-fit/openff/bespokefit/workflows/bespoke.py in optimization_schema_from_molecule(self, molecule, index)
394         # the parent molecule is put in canonical order by fragmenter to use the new parent
395         attributes = MoleculeAttributes.from_openff_molecule(
--> 396             molecule=fragment_data[0].parent_molecule
397         )
398         molecule_schema = MoleculeSchema(
IndexError: list index out of range

```

I'm guessing perhaps adding a condition to check for fragments after calling the fragmentation_engine and if none generated to directly proceed with the parent molecule could fix this.

### PR DESCRIPTION
**Pull request by jthorton**
## Description
This PR changes how bespoke smirks are generated for fragment molecules as the current method would sometimes produce patterns that would not correctly transfer back to the parent molecule. Now the targeted parameters are grouped by molecule symmetry and a cluster graph is used to generate a smirks pattern that captures the symmetry equivalent terms in both the fragment and parent molecule. 

Fixes #40 

## Status
- [x] Ready to go

### CODE REVIEW

File: openff/bespokefit/bespoke/smirks.py [hunk 1]
```diff
@@ -158,148 +180,232 @@ def _get_all_bespoke_smirks(
         values.
         """
         bespoke_smirks = []
+        fragment_is_parent = True
+        fragment_molecule = fragment_data.fragment_molecule
+        # check if the fragment is the same as the parent, do it once for speed
+        if fragment_data.parent_molecule != fragment_molecule:
+            fragment_is_parent = False
```
Comment (SimonBoothroyd) [inline openff/bespokefit/bespoke/smirks.py:187]:
```suggestion
        fragment_molecule = fragment_data.fragment_molecule
        # check if the fragment is the same as the parent, do it once for speed
        fragment_is_parent = fragment_data.parent_molecule == fragment_molecule
```

File: openff/bespokefit/bespoke/smirks.py [hunk 2]
```diff
@@ -158,148 +180,232 @@ def _get_all_bespoke_smirks(
...
+            for angle in fragment_mol.angles
+        ]
+        angle_groups = SmirksGenerator.group_valence_by_symmetry(fragment_mol, angles)
+        angle_groups = [*angle_groups.values()]
+
+        for angle_group in angle_groups:
+            target_atoms = [angle_group]
+            target_molecules = [fragment_data.parent_molecule]
+            if not fragment_is_parent:
+                parent_angles = [
```
Comment (SimonBoothroyd) [inline openff/bespokefit/bespoke/smirks.py:321]: Could you add some tests to make sure the logic in the `if not fragment_is_parent:` conditionals is tested?

**Review by SimonBoothroyd — APPROVED** [review]
Thanks for putting this together @jthorton - just a few minor suggestions but otherwise LGTM!

File: openff/bespokefit/fragmentation/model.py [hunk 3]
```diff
@@ -138,6 +143,52 @@ def build_fragment_data(cls, result: "FragmentationResult") -> List[FragmentData
...
+        from openff.qcsubmit.workflow_components import ScanEnumerator, ScanFilter
+
+        logger.warning(
+            f"The molecule {molecule} could not be fragmented as the stereochemistry of the fragment could not be "
+            f"made to match the parent. The parent will be used as the target molecule."
+        )
+
+        scanner = ScanEnumerator()
+        # tag all possible scans
+        scanner.add_torsion_scan("[*:1]~[!$(*#*)&!D1:2]-,=;!@[!$(*#*)&!D1:3]~[*:4]")
```
[the remark this answers is not in the document]
Comment (jthorton) [reply openff/bespokefit/fragmentation/model.py:163]: Ah, that's perfect thanks!

File: openff/bespokefit/bespoke/smirks.py (hunk 1 above)
Comment (jthorton) [reply openff/bespokefit/bespoke/smirks.py:187]: Nice, I would never have thought of that!

File: openff/bespokefit/fragmentation/fragmenter.py [hunk 4]
```diff
@@ -65,13 +64,11 @@ def fragment(self, molecule: Molecule) -> List[FragmentData]:
...
             result = fragment_factory.fragment(molecule=molecule)
             fragments = self.build_fragment_data(result=result)
-            return fragments
 
         except RuntimeError:
-            raise FragmenterError(
-                f"The molecule {molecule} could not be fragmented so no fitting target "
-                f"was made."
-            )
+            fragments = self.build_fragments_from_parent(molecule=molecule)
```
[the remark this answers is not in the document]
Comment (jthorton) [reply openff/bespokefit/fragmentation/fragmenter.py:69]: Good idea added to the start of the `build_fragments_from_parent` function.

File: openff/bespokefit/bespoke/smirks.py (hunk 2 above)
Comment (jthorton) [reply openff/bespokefit/bespoke/smirks.py:321]: Great shout, this also uncovered a few bugs caused by typos so thanks for this!

**Pull request merged**